src/audio_call.c

```diff
@@ -386,76 +413,85 @@ void callback_recv_invite(Tox *m, uint32_t friend_number)
+            windows[i]->onInvite(windows[i], CallControl.av, friend_number, call->state);
         }
     }
 }
 void callback_recv_ringing(uint32_t friend_number)
 {
+    const Call *call = &CallControl.calls[friend_number];
+
     for (uint8_t i = 0; i < MAX_WINDOWS_NUM; ++i) {
         if (windows[i] != NULL && windows[i]->onRinging != NULL && windows[i]->num == friend_number) {
-            windows[i]->onRinging(windows[i], CallControl.av, friend_number, CallControl.call_state);
+            windows[i]->onRinging(windows[i], CallControl.av, friend_number, call->state);
         }
     }
 }
 void callback_recv_starting(uint32_t friend_number)
 {
+    Call *call = &CallControl.calls[friend_number];
+
     for (uint8_t i = 0; i < MAX_WINDOWS_NUM; ++i) {
         if (windows[i] != NULL && windows[i]->onStarting != NULL && windows[i]->num == friend_number) {
-            windows[i]->onStarting(windows[i], CallControl.av, friend_number, CallControl.call_state);
-
-            if (0 != start_transmission(windows[i], &CallControl.calls[friend_number])) { /* YEAH! */
-                line_info_add(windows[i], NULL, NULL, NULL, SYS_MSG, 0, 0, "Error starting transmission!");
-            }
+            windows[i]->onStarting(windows[i], CallControl.av, friend_number, call->state);
 
-            return;
+            start_call(windows[i], call);
         }
     }
 }
 void callback_recv_ending(uint32_t friend_number)
 {
+    const Call *call = &CallControl.calls[friend_number];
+
     for (uint8_t i = 0; i < MAX_WINDOWS_NUM; ++i) {
         if (windows[i] != NULL && windows[i]->onEnding != NULL && windows[i]->num == friend_number) {
-            windows[i]->onEnding(windows[i], CallControl.av, friend_number, CallControl.call_state);
+            windows[i]->onEnding(windows[i], CallControl.av, friend_number, call->state);
         }
     }
 }
 void callback_call_started(uint32_t friend_number)
 {
+    Call *call = &CallControl.calls[friend_number];
+
     for (uint8_t i = 0; i < MAX_WINDOWS_NUM; ++i) {
         if (windows[i] != NULL && windows[i]->onStart != NULL && windows[i]->num == friend_number) {
-            windows[i]->onStart(windows[i], CallControl.av, friend_number, CallControl.call_state);
+            windows[i]->onStart(windows[i], CallControl.av, friend_number, call->state);
 
-            if (0 != start_transmission(windows[i], &CallControl.calls[friend_number])) {  /* YEAH! */
-                line_info_add(windows[i], NULL, NULL, NULL, SYS_MSG, 0, 0, "Error starting transmission!");
-                return;
-            }
+            start_call(windows[i], call);
         }
     }
 }
 void callback_call_canceled(uint32_t friend_number)
 {
+    const Call *call = &CallControl.calls[friend_number];
+
     for (uint8_t i = 0; i < MAX_WINDOWS_NUM; ++i) {
         if (windows[i] != NULL && windows[i]->onCancel != NULL && windows[i]->num == friend_number) {
```

Comment:
I feel like we should factor out this long a && b && c expression into a function and use it in these callbacks. WDYT?

Reply:
Ah, they are all slightly different. Maybe factor out the 2 common bits?

Reply:
On third read, it seems you can generalise these functions entirely and make each of them 1 line.

Reply:
Can't generalise it after all (I misread things), and factoring out the common bits seems less useful now. Never mind.